### PR TITLE
Improve docs for custom events

### LINKED ISSUE #1555: Filters don't work on custom events
`CallableEventWithFilter` does not work with custom events:

```python
from ignite.engine import Engine, EventEnum
    
def train_step(engine, batch):
    engine.fire_event(TestEvents.TEST_EVENT)
engine = Engine(train_step)

class TestEvents(EventEnum):
    TEST_EVENT = 'test_event'
engine.register_events(*TestEvents)

@engine.on(TestEvents.TEST_EVENT(every=100))
def foo(engine):
    print("hello world")

def forever():
    while True:
        yield 0.
    
engine.run(forever(), max_epochs=10, epoch_length=1)
```

The error is:
```
RuntimeError: Unknown event name '<event=TEST_EVENT, filter=<function CallableEventWithFilter.every_event_filter.<locals>.wrapper at 0x7fb7d1244710>>'
```

### PR DESCRIPTION
**Pull request by jrieke**
Fixes #1555 

Description: Improved [the docs](https://pytorch.org/ignite/concepts.html#events-and-handlers) on how to create custom events according to the example in #1555.

Check list:
* [ ] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

File: docs/source/concepts.rst [hunk 1]
```diff
@@ -254,6 +260,7 @@ These events could be used to attach any handler and are fired using :meth:`~ign
     @engine.on(Events.STARTED)
     def fire_custom_events(engine):
          engine.fire_event(CustomEvents.CUSTOM_STARTED)
+         engine.state.custom_started += 1  # increase state counter for this event (required)
```
Comment (vfdev-5) [inline docs/source/concepts.rst:263]: Let's say this is required for filtering only and can be omitted if no filtering is done. And another note to add is that this is a temporary solution and subject to change in future releases.

File: docs/source/concepts.rst [hunk 2]
```diff
@@ -241,9 +241,15 @@ and be registered with :meth:`~ignite.engine.engine.Engine.register_events` in a
         CUSTOM_STARTED = 'custom_started'
         CUSTOM_COMPLETED = 'custom_completed'
 
-    engine.register_events(*CustomEvents)
+    engine.register_events(
+        *CustomEvents,
+        event_to_attr={
```
Comment (vfdev-5) [inline docs/source/concepts.rst:246]: This part is not required actually. I'd put everything in the following way:
- simple custom events without filtering (current doc)
- custom events with filtering:
   - engine.state should have corresponding attributes
   - event_to_attr is required when `register_events`
   - should increase counters (subject to be removed)

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @jrieke !
I left few comments to improve things

**Comment by jrieke** [comment]
@vfdev-5 Ok I restructured it as you suggested. I removed all changes from the original docs and added the 3 steps for filtering as a paragraph below it.

File: docs/source/concepts.rst [hunk 3]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
...
+.. code-block:: python
+
+    event_to_attr={
+            CustomEvents.CUSTOM_STARTED: "custom_started",
+            CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+        }
+
+- you should increase the counter for the event whenever you fire the event, e.g. `engine.state.custom_started += 1`
+
+.. Note ::
```
Comment (vfdev-5) [inline docs/source/concepts.rst:277]:
```suggestion
.. warning::
```

**Review by vfdev-5 — COMMENTED** [review]
A minor fix to add, otherwise it is OK. Thanks @jrieke !

File: docs/source/concepts.rst [hunk 4]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
...
+
+- `engine.state` should have corresponding attributes for the events, e.g. `engine.state.custom_started`
+- you need to pass a dict `event_to_attr` to `register_events`, which maps between events and state attributes, e.g.
+
+.. code-block:: python
+
+    event_to_attr={
+            CustomEvents.CUSTOM_STARTED: "custom_started",
+            CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+        }
```
Comment (vfdev-5) [inline docs/source/concepts.rst:273]:
```suggestion
    event_to_attr = {
        CustomEvents.CUSTOM_STARTED: "custom_started",
        CustomEvents.CUSTOM_COMPLETED: "custom_completed",
    }
```

File: docs/source/concepts.rst [hunk 5]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
+
+- `engine.state` should have corresponding attributes for the events, e.g. `engine.state.custom_started`
```
Comment (vfdev-5) [inline docs/source/concepts.rst:265]:
```suggestion
- ``engine.state`` should have corresponding attributes for the events, e.g. ``engine.state.custom_started``
```

File: docs/source/concepts.rst [hunk 6]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
...
+- you need to pass a dict `event_to_attr` to `register_events`, which maps between events and state attributes, e.g.
```
Comment (vfdev-5) [inline docs/source/concepts.rst:266]:
```suggestion
- you need to pass a dict `event_to_attr` to :meth:`~ignite.engine.engine.Engine.register_events`, which maps between events and state attributes, e.g.
```

File: docs/source/concepts.rst [hunk 7]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
...
+
+.. code-block:: python
+
+    event_to_attr={
+            CustomEvents.CUSTOM_STARTED: "custom_started",
+            CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+        }
+
+- you should increase the counter for the event whenever you fire the event, e.g. `engine.state.custom_started += 1`
```
Comment (vfdev-5) [inline docs/source/concepts.rst:275]:
```suggestion
- you should increase the counter for the event whenever you fire the event, e.g. ``engine.state.custom_started += 1``
```

File: docs/source/concepts.rst [hunk 8]
```diff
@@ -260,6 +260,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with `CallableEventWithFilter`), you need to do 3 more things:
```
Comment (vfdev-5) [inline docs/source/concepts.rst:263]:
```suggestion
If you want to use filtering with custom events (e.g. with ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```

Comment (vfdev-5) [inline docs/source/concepts.rst:263]: Can you also please add a subsection to the line 229 : 

    Custom events
    ``````````````

**Review by vfdev-5 — COMMENTED** [review]
Few more comments

File: docs/source/concepts.rst (hunk 8 above)
Comment (jrieke) [reply docs/source/concepts.rst]: Done!

File: docs/source/concepts.rst [hunk 9]
```diff
@@ -260,6 +263,25 @@ These events could be used to attach any handler and are fired using :meth:`~ign
    See the source code of :class:`~ignite.contrib.engines.create_supervised_tbptt_trainer` for an example of usage of
    custom events.
 
+If you want to use filtering with custom events (e.g. with ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```
Comment (vfdev-5) [inline docs/source/concepts.rst:266]:
```suggestion
If you want to use filtering with custom events (e.g. ``CustomEvents.CUSTOM_STARTED(every=5)``), you need to do 3 more things:
```

**Review by vfdev-5 — COMMENTED** [review]
LGTM, thanks @jrieke !

**Review by vfdev-5 — APPROVED** [review]

**Pull request merged**